docker_images/nemo/app/pipelines/automatic_speech_recognition.py

```diff
@@ -0,0 +1,74 @@
+from typing import Dict
+
+import os
+import librosa
+import numpy as np
+import uuid
+import tempfile
+import soundfile
+
+from app.pipelines import Pipeline
+from huggingface_hub import hf_hub_download
+from huggingface_hub.hf_api import HfFolder
+
+
+import nemo.collections.asr as nemo_asr
+
+
+class AutomaticSpeechRecognitionPipeline(Pipeline):
+    def __init__(self, model_id: str):
+        # IMPLEMENT_THIS
+        # Preload all the elements you are going to need at inference.
+        # For instance your model, processors, tokenizer that might be needed.
+        # This function is only called once, so do all the heavy processing I/O here
+
+        # Precheck for API key 
+        is_token_available = HfFolder.get_token() is not None
+
+        # Prepare file name from model_id
+        filename = model_id.split("/")[-1] + ".nemo"
+        path = hf_hub_download(repo_id=model_id, filename=filename, use_auth_token=is_token_available)
+
+        # Load model
+        self.model = nemo_asr.models.ASRModel.restore_from(path)
+        self.model.freeze()
+        
+        # Pre-Initialize RNNT decoding strategy
+        if hasattr(self.model, 'change_decoding_strategy'):
+            model.change_decoding_strategy(None)
+
+        # IMPLEMENT_THIS : Please define a `self.sampling_rate` for this pipeline
+        # to automatically read the input correctly
+        self.sampling_rate = self.model.cfg.sample_rate
+        
+
+    def __call__(self, inputs: np.array) -> Dict[str, str]:
+        """
+        Args:
+            inputs (:obj:`np.array`):
+                The raw waveform of audio received. By default at self.sampling_rate, otherwise 16KHz.
+        Return:
+            A :obj:`dict`:. The object return should be liked {"text": "XXX"} containing
+            the detected langage from the input audio
+        """
+        inputs = self.process_audio_file(inputs)
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            audio_path = os.path.join(tmpdir, f'audio_{uuid.uuid4()}.wav')
+            soundfile.write(audio_path, inputs, self.sampling_rate)
+
+            transcriptions = self.model.transcribe([audio_path])
```

Comment:
The NeMo ASR pipeline is quite involved, and we have a layered API to perform inference. transcribe() kind of generalizes the API across various scopes (5 different encoders x {ctc decoder, rnnt decoder} x {char, subword} x languages).  Since it is designed for offline transcription of files, it somewhat depends on the assumption that file paths are provided so that the data loader can be reused per model. 

The audio is stored temporarily (to reduce data storage concerns) and is immediately deleted after inference. Its definitely a performance hit, perhaps later nemo versions can have higher level APIs that deal with audio directly. Once we stabilize such an API, we can update here. Does that seem reasonable?

Reply:
Very reasonable ! what I suggest is we keep the code as-is, we check real world performance. If it's not too bad we keep it as-is, if it is I will enable some kind of `tmpfs`.

When there's better code we can update as desired.